libcudacxx/include/cuda/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,310 @@
+
+    // CUDA drivers before 11.5 have known incompatibilities with the async allocator.
+    // We'll disable `cudaMemPoolReuseAllowOpportunistic` if cuda driver < 11.5.
+    // See https://github.com/NVIDIA/spark-rapids/issues/4710.
+    int __driver_version = 0;
+    _CCCL_TRY_CUDA_API(::cudaDriverGetVersion, "Failed to call cudaDriverGetVersion", &__driver_version);
+
+    constexpr int __min_async_version = 11050;
+    if (__driver_version < __min_async_version)
+    {
+      int __disable_reuse = 0;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemPoolSetAttribute,
+        "Failed to call cudaMemPoolGetAttribute",
+        __cuda_pool_handle,
+        ::cudaMemPoolReuseAllowOpportunistic,
+        &__disable_reuse);
+    }
+
+    _CCCL_TRY_CUDA_API(
+      ::cudaMemPoolSetAttribute,
+      "Failed to call cudaMemPoolReuseAllowOpportunistic",
+      __cuda_pool_handle,
+      ::cudaMemPoolAttrReleaseThreshold,
+      &__release_threshold);
+
+    // allocate the requested initial size to pprime the pool
+    if (__initial_pool_size != 0)
+    {
+      void* __ptr{nullptr};
+      _CCCL_TRY_CUDA_API(
+        ::cudaMallocAsync,
+        "cuda_memory_pool failed allocate the initial pool size",
+        &__ptr,
+        __initial_pool_size,
+        ::cudaStream_t{0});
+
+      _CCCL_ASSERT_CUDA_API(
+        ::cudaFreeAsync, "cuda_memory_pool failed to free the initial pool allocation", __ptr, ::cudaStream_t{0});
+    }
+    return __cuda_pool_handle;
+  }
+
+public:
+  /**
+   * @brief Constructs a cuda_memory_pool with the optionally specified initial pool size
+   * and release threshold.
+   *
+   * If the pool size grows beyond the release threshold, unused memory held by the pool will be
+   * released at the next synchronization event.
+   *
+   * @throws ::cuda::cuda_error if the CUDA version does not support `cudaMallocAsync`
+   *
+   * @param initial_pool_size Optional initial size in bytes of the pool. If no value is provided,
+   * initial pool size is half of the available GPU memory.
+   * @param release_threshold Optional release threshold size in bytes of the pool. If no value is
+   * provided, the release threshold is set to the total amount of memory on the current device.
+   */
+  cuda_memory_pool(
+    const size_t initial_pool_size                             = 0,
+    const size_t release_threshold                             = 0,
+    const cudaMemAllocationHandleType __allocation_handle_type = cudaMemAllocationHandleType::cudaMemHandleTypeNone)
+      : __pool_handle_(__create_cuda_mempool(initial_pool_size, release_threshold, __allocation_handle_type))
+  {}
+
+  /**
+   * @brief Disables construction from a plain `cudaMemPool_t`. We want to ensure clean ownership semantics
+   */
+  cuda_memory_pool(::cudaMemPool_t) = delete;
+
+  cuda_memory_pool(cuda_memory_pool const&)            = delete;
+  cuda_memory_pool(cuda_memory_pool&&)                 = delete;
```

Comment:
Is this solely because you want to avoid empty states in this class?

Reply:
Yes. This should always contain a valid cudaMemPool_t

Reply:
That said, I am open to allow move operations